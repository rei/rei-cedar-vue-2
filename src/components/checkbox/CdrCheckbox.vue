<template>
  <!-- eslint-disable max-len -->
  <div :class="$style['cdr-checkbox__wrap']">
    <label
      :class="[modifierClass, labelClass]"
      ref="label"
    >
      <input
        :class="[$style['cdr-checkbox__input'], inputClass]"
        type="checkbox"
        v-bind="$attrs"
        v-model="newValue"
        :true-value="customValue ? null : trueValue"
        :false-value="customValue ? null : falseValue"
        :value="customValue"
        :indeterminate="indeterminate"
        @change="updateValue(newValue, $event)"
        ref="checkbox"
      >
      <span :class="$style['cdr-checkbox__figure']" />
      <div :class="[$style['cdr-checkbox__content'], contentClass]">
        <!-- @slot innerHTML inside of checkbox component -->
        <slot/>
      </div>
    </label>
  </div>
</template>

<script>
import modifier from 'mixinsdir/modifier';

/**
 * Cedar 2 component for checkbox
 * **NOTE:** `v-model` is required.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrCheckbox',
  mixins: [modifier],
  inheritAttrs: false,
  props: {
    /**
     * Class that is added to the label for custom styles
     */
    labelClass: String,
    /**
     * Class that is added to the input for custom styles
     */
    inputClass: String,
    /**
     * Class that is added to the slot wrapper for custom styles
     */
    contentClass: String,
    /**
     * Show checkbox in indeterminate state. (NOTE: this is a visual-only state and there is no logic for when to show it)
    */
    indeterminate: {
      type: Boolean,
      default: false,
    },
    /**
     * The value when checked.
    */
    trueValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      default: true,
    },
    /**
     * The value when unchecked.
    */
    falseValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      default: false,
    },
    /**
     * The value when used in a checkbox group. Replaces `trueValue` and `falseValue`.
    */
    customValue: [String, Number, Boolean, Object, Array, Symbol, Function],
    /** @ignore */
    value: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
    },
  },
  data() {
    return {
      newValue: this.value,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-checkbox';
    },
  },
  watch: {
    value(val) {
      this.newValue = val;
    },
    newValue(val) {
      /**
       * `v-model` value. Fires on check/uncheck.
       * @event input
       * @type value | event
       * */
      this.$emit('input', val);
    },
  },
  methods: {
    updateValue(newValue, e) {
      this.newValue = newValue;
      this.$emit('change', newValue, e);
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/CdrCheckbox.pcss';
</style>
