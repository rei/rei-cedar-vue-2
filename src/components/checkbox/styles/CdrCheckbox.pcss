/* ==========================================================================
  # CdrCheckbox
  
  All values should map to variables in
  vars/CdrCheckbox.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Checkbox
      :Elements
        :wrap
        :input
        :figure
      :States
      :Modifiers
        :compact
        :hide-figure
========================================================================== */

.cdr-checkbox {
  @include redwood-display-30;

  color: $checkbox-label-color;
  display: inline-block;
  margin: 0;
  position: relative;
  padding-left: calc($checkbox-figure-size + $checkbox-figure-space-right);

  /* Elements
    ========================================================================== */

  /* Wrap
  ========== */
  &__wrap {
    /* get rid of inline-block whitespace from label element */
    font-size: 0;
  }

  /* Input
  ========== */
  &__input {
    position: absolute;
    opacity: 0;
    z-index: -1;
  }

  /* Figure
  ========== */
  &__figure {
    position: absolute;
    top: $checkbox-figure-space-top;
    left: 0;
    width: $checkbox-figure-size;
    height: $checkbox-figure-size;
    border-radius: $checkbox-figure-radius;
    background-color: $checkbox-figure-background-color;
    border: 1px solid $checkbox-figure-border-color;
  }

  /* States
    ========================================================================== */

  /* Checked
  ========== */
  &__input:checked ~ &__figure {
    border-color: $checkbox-figure-checked-border-color;
    background-color: $checkbox-figure-checked-background-color;
    background-image: svg-load('../../../assets/icons/rei/check-lg.svg', fill=$(checkbox-figure-icon-fill));
    background-repeat: no-repeat;
  }

  /* Indeterminate
  ========== */
  &__input[indeterminate] ~ .cdr-checkbox__figure {
    border-color: $checkbox-figure-indeterminate-border-color;
    
    &::after {
      content: '';
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
      width: $checkbox-figure-indeterminate-size;
      height: $checkbox-figure-indeterminate-size;
      background-color: $checkbox-figure-indeterminate-background-color;
      border-radius: $checkbox-figure-indeterminate-border-radius;
    }
  }

  /* Hover
  ========== */
  &:hover {
    & > .cdr-checkbox__figure {
      cursor: pointer;
      border-color: $checkbox-figure-hover-border-color;
      background-color: $checkbox-figure-hover-background-color;
    }

    & > .cdr-checkbox__content {
      cursor: pointer;
    }

    & > .cdr-checkbox__input {
      /* Hover + Checked
      ========== */
      &:checked ~ .cdr-checkbox__figure {
        border-color: $checkbox-figure-checked-hover-border-color;
        background-color: $checkbox-figure-checked-hover-background-color;
      }

      /* Hover + Indeterminate
      ========== */
      &[indeterminate] ~ .cdr-checkbox__figure {
        border-color: $checkbox-figure-checked-hover-border-color;

        &::after {
          background-color: $checkbox-figure-checked-hover-background-color;
        }
      }
    }
  }

  /* Disabled
  ========== */
  &__input:disabled {
    & ~ .cdr-checkbox__content {
      color: $checkbox-label-disabled-color !important;
      cursor: default;
    }

    & ~ .cdr-checkbox__figure {
      border-color: $checkbox-figure-disabled-border-color !important;
      background-color: $checkbox-figure-disabled-background-color !important;
      cursor: default;
    }

    /* Disabled + Checked
    ========== */
    &:checked {
      & ~ .cdr-checkbox__figure {
        background-color: $checkbox-figure-checked-disabled-background-color !important;
      }
    }

    /* Disabled + Indeterminate
    ========== */
    &[indeterminate] {
      & ~ .cdr-checkbox__figure {
        border-color: $checkbox-figure-disabled-border-color !important;
        
        &::after {
          background-color: $checkbox-figure-checked-disabled-background-color !important;
        }
      }
    }
  }

  /* Active
  ========== */
  &:active > .cdr-checkbox__figure {
    border-color: $checkbox-figure-active-border-color;
  }

  /* Focus
  ========== */
  &__input:focus ~ .cdr-checkbox__figure {
    border-color: $checkbox-figure-focus-border-color;
    box-shadow: $checkbox-focus-outline;
  }

  /* Modifiers
    ========================================================================== */

  /* Compact
    ========== */

  &--compact {
    @include redwood-display-20;

    & > .cdr-checkbox__figure {
      top: $checkbox-compact-figure-space-top;
    }
  }

  /* Hide Figure
    ========== */

  &--hide-figure {
    padding-left: 0;

    & > .cdr-checkbox__figure {
      display: none;
    }
  }
}
