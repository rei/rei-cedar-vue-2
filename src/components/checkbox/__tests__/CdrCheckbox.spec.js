import { shallowMount, mount } from '@vue/test-utils';
import CdrCheckbox from 'componentdir/checkbox/CdrCheckbox';

describe('CdrCheckbox', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrCheckbox, {
      propsData: {
        labelClass: 'custom-label-class',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('is type checkbox', () => {
    const wrapper = shallowMount(CdrCheckbox);
    expect(wrapper.vm.$refs.checkbox.hasAttribute('type', 'checkbox')).toBe(true);
  });

  it('renders a label element', () => {
    const wrapper = shallowMount(CdrCheckbox);
    expect(wrapper.vm.$refs.label.tagName).toBe('LABEL');
  });
  
  it('adds a custom label class correctly', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        labelClass: 'custom-label-class',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.vm.$refs.label.classList.contains('custom-label-class')).toBe(true);
  });
  
  it('adds a custom input class correctly', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        inputClass: 'custom-input-class',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.vm.$refs.checkbox.classList.contains('custom-input-class')).toBe(true);
  });
  
  it('adds a custom content class correctly', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        contentClass: 'custom-content-class',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.find('.custom-content-class').exists()).toBe(true);
  });

  it('watches values correctly', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        value: false,
      },
    });
    wrapper.setProps({ value: true });
    expect(wrapper.vm.value).toBe(true);
    expect(wrapper.vm.newValue).toBe(true);
  });

  it('emits change events with correct values for default checkbox', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        value: false,
      },
    });
    const cb = wrapper.find({ ref: 'checkbox'});
    cb.trigger('click');
    expect(wrapper.emitted().change[0][0]).toBe(true);
    cb.trigger('click');
    expect(wrapper.emitted().change[1][0]).toBe(false);
  });
  
  it('emits change events with correct values for custom checkbox', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        trueValue: 'checked',
        falseValue: 'unchecked',
        value: '',
      },
    });
    const cb = wrapper.find({ ref: 'checkbox'});
    cb.trigger('click');
    expect(wrapper.emitted().change[0][0]).toBe('checked');
    cb.trigger('click');
    expect(wrapper.emitted().change[1][0]).toBe('unchecked');
  });
  
  it('emits change events with correct values for group checkbox', () => {
    const wrapper = shallowMount(CdrCheckbox, {
      propsData: {
        customValue: 'b',
        value: ['a'],
      },
    });
    const cb = wrapper.find({ ref: 'checkbox'});
    cb.trigger('click');
    expect(wrapper.emitted().change[0][0]).toEqual(['a', 'b']);
    cb.trigger('click');
    expect(wrapper.emitted().change[1][0]).toEqual(['a']);
  });
});
