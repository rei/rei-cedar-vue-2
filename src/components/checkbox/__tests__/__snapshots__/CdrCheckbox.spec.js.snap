// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrCheckbox renders correctly 1`] = `
<div
  class="cdr-checkbox__wrap"
>
  <label
    class="cdr-checkbox custom-label-class"
  >
    <input
      class="cdr-checkbox__input"
      type="checkbox"
    />
    <span
      class="cdr-checkbox__figure"
    />
    <div
      class="cdr-checkbox__content"
    >
      Label Test
    </div>
  </label>
</div>
`;
